### PR TITLE
feat: Image format safety — magic byte validation + HEIC/AVIF/JXL support

### LINKED ISSUE #52: "Bad" images cause segfault
We noticed that sometimes people uploaded a weird image that libgd did not understand. The runtime then segfaulted. IIRC at least one of these was a TIFF file.

### PR DESCRIPTION
**Pull request by penguinboi**
## Summary

- **Fixes #52** — Unsupported image uploads (HEIC, AVIF, JXL) no longer crash the server via GD segfault. Magic byte detection identifies the format before touching GD.
- **Addresses #535** — HEIC, AVIF, and JXL uploads are now converted to JPEG server-side via CLI tools (`heif-convert`, `djxl`) before GD processing.
- Unknown formats get a descriptive 400 error instead of a segfault.

## What changed

| File | Change |
|------|--------|
| `Sources/swiftarr/Image/ImageFormatDetector.swift` | New — detects image format from magic bytes (JPEG, PNG, GIF, WebP, TIFF, BMP, HEIC, AVIF, JXL) |
| `Sources/swiftarr/Image/ImageFormatConverter.swift` | New — converts non-GD formats to JPEG via CLI tools with temp files, timeout, cleanup |
| `Sources/swiftarr/Protocols/ImageHandler.swift` | Modified — `loadImageFromData()` now detects format first, routes to GD or converter, falls back to TGA/WBMP trial-parse for obscure formats |
| `scripts/init-prereqs.sh` | Added `libheif-examples libjxl-tools` to Docker runtime apt install |
| `Tests/AppTests/ImageFormatDetectorTests.swift` | 19 unit tests for all format detections + edge cases |
| `Tests/AppTests/ImageFormatConverterTests.swift` | 4 tests (2 error paths, 1 HEIC conversion with skip, 1 temp cleanup) |

## Design decisions

- **CLI tools over native libraries** — `heif-convert` and `djxl` are simpler to add, review, and maintain. All known clients already convert before upload, so this is a safety net — subprocess overhead is fine.
- **Magic bytes replace blind parsing** — Instead of trying all 7 GD parsers and catching errors, we identify the format from header bytes and use the right parser directly. TGA/WBMP fall back to trial-parse (no reliable magic bytes).
- **JPEG quality 95 intermediate** — Conversion output is Q95 because GD re-encodes at Q90 on final export. Two rounds of JPEG, so we keep intermediate quality high.
- **30-second subprocess timeout** — Safety valve for malformed files.

## Test plan

- [x] 19 detector unit tests pass (all formats + edge cases)
- [x] 4 converter unit tests pass (1 skips when `heif-convert` not installed — expected)
- [x] Full project builds clean
- [ ] Docker build with converter packages (needs Ubuntu 24.04 environment)
- [ ] Manual test: upload a HEIC photo via API endpoint
- [ ] Manual test: upload random bytes, verify 400 error

### CODE REVIEW

File: docs/superpowers/plans/2026-04-02-image-format-safety.md [hunk 1]
```diff
@@ -0,0 +1,870 @@
+# Image Format Safety Implementation Plan
```
Comment (hendricksond) [inline docs/superpowers/plans/2026-04-02-image-format-safety.md]: I don't think we need the `docs/superpowers` or `CLAUDE.md` files checked in, though I'd be happy to have more discussion about that if others think they should be here!

Comment (penguinboi) [reply docs/superpowers/plans/2026-04-02-image-format-safety.md]: Opps! yes I agree. let me get that out of there. At least temporarily you can see the logic of my proposed change.

Comment (cohoe) [reply docs/superpowers/plans/2026-04-02-image-format-safety.md]: I agree with `docs/superpowers`. I have 10 million Cursor Plans myself that I do not feel carry significant value after their implementation and only serve to be out-of-date references and cost input tokens. I am OK with a `CLAUDE.md` with the following stipulations:

1) Do not duplicate the environment setup. We already have that in `Docs`.
2) We don't need the team listed, especially since what's in there is not entirely accurate.
3) Don't need to list remotes. Github already has a method to view forks, and we don't care who has forks.
4) Eliminate Pending Contributions or other "churn" in the document based on whatever the current context/project is. Code notes are fine but honestly I would prefer those under `Docs` instead (kinda like what Tricordarr does). Claude-specific notes, fine. But basically I don't want `CLAUDE.md` to receive changes based on every single PR.

Comment (hendricksond) [reply docs/superpowers/plans/2026-04-02-image-format-safety.md]: Eventually, we'll probably want a `CLAUDE.md` and/or `AGENTS.md` file, but that should really be its own PR 😁

Comment (cohoe) [reply docs/superpowers/plans/2026-04-02-image-format-safety.md]: ^^ what he said

Comment (penguinboi) [reply docs/superpowers/plans/2026-04-02-image-format-safety.md]: both files have been removed from the PR

File: Sources/swiftarr/Image/ImageFormatDetector.swift [hunk 2]
```diff
@@ -0,0 +1,131 @@
+// ABOUTME: Identifies image formats by reading magic bytes from file headers.
```
Comment (cohoe) [inline Sources/swiftarr/Image/ImageFormatDetector.swift]: nit: `ABOUTME` is kinda silly. Just comment like normal.

File: Sources/swiftarr/Image/ImageFormatDetector.swift [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+	var needsConversion: Bool {
+		switch self {
+		case .heic, .avif, .jxl: return true
+		default: return false
+		}
+	}
+}
+
+/// Detects image format from file header bytes.
+struct ImageFormatDetector {
```
Comment (cohoe) [inline Sources/swiftarr/Image/ImageFormatDetector.swift:38]: We should leverage a module for this. https://github.com/velocityzen/FileType

File: Sources/swiftarr/Protocols/ImageHandler.swift [hunk 4]
```diff
@@ -66,25 +66,50 @@ func getImagePath(for image: String, format: String? = nil, usage: ImageUsage, s
...
-		for type in imageTypes {
-			foundImage = try? GDImage(data: data, as: type) 
-			if foundImage != nil{
-				foundType = type
-				break
+		let detected = ImageFormatDetector.detect(data)
+		let imageData: Data
+		let imageType: ImportableFormat
+
+		if let gdFormat = detected.gdFormat {
```
Comment (cohoe) [inline Sources/swiftarr/Protocols/ImageHandler.swift:78]: On boat, @challfry had mentioned some interest in replacing our hacky GD lib with something more real. https://github.com/libvips/libvips looks like it may be a compatible and more better library. Typically we develop on MacOS (ARM64) but run on Linux (AMD64) so whatever solution has to support both. I think it would be good to explore this as a native library rather than farm out to shell commands.

If the CLI pattern is determined to be "more better" the appropriate packages/dependencies need to be added to the setup docs for both MacOS and Linux.

File: Sources/swiftarr/Image/ImageFormatConverter.swift [hunk 5]
```diff
@@ -0,0 +1,148 @@
+// ABOUTME: Converts image formats that GD cannot handle (HEIC, AVIF, JXL) to JPEG.
+// ABOUTME: Shells out to CLI tools (heif-convert, djxl) with temp files and timeout.
```
Comment (cohoe) [inline Sources/swiftarr/Image/ImageFormatConverter.swift]: nit: `ABOUTME` is kinda silly. Just have the comment.

File: Sources/swiftarr/Image/ImageFormatConverter.swift [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+		case .conversionFailed(let reason):
+			return "Image conversion failed: \(reason)"
+		case .timeout:
+			return "Image conversion timed out"
+		}
+	}
+}
+
+/// Converts non-GD image formats to JPEG via CLI tools.
+struct ImageFormatConverter {
```
Comment (cohoe) [inline Sources/swiftarr/Image/ImageFormatConverter.swift:30]: See above/below re: philosophy around potential library replacement.

**Review by cohoe — CHANGES_REQUESTED** [review]

File: Sources/swiftarr/Protocols/ImageHandler.swift (hunk 4 above)
Comment (cohoe) [reply Sources/swiftarr/Protocols/ImageHandler.swift:78]: Library may also implicitly solve #541

**Comment by penguinboi** [comment]
Removed the ABOUTME prefixes from the file comments.

Re: libvips — I think that's a solid direction, especially since it could address #541 too. Before I go further on this PR, wanted to check: should I pivot to a libvips-based approach, or is the CLI version OK to land as-is while we explore libvips as a follow-up?

Happy to go either way — just don't want to polish the CLI path if we're going to replace it.

**Comment by cohoe** [comment]
> Removed the ABOUTME prefixes from the file comments.
> 
> Re: libvips — I think that's a solid direction, especially since it could address #541 too. Before I go further on this PR, wanted to check: should I pivot to a libvips-based approach, or is the CLI version OK to land as-is while we explore libvips as a follow-up?
> 
> Happy to go either way — just don't want to polish the CLI path if we're going to replace it.

Lets do the libvips approach. If that proves to be untenable or more trouble than it's worth then we can revisit the CLI as implemented here.

**Comment by penguinboi** [comment]
Superseded by #548 — pivoted to full libgd→libvips replacement per review feedback. This PR's CLI-based approach is preserved on the branch as a fallback if libvips proves untenable.

**Comment by penguinboi** [comment]
Closing this in favor of #548, which has been merged and is the canonical fix for #535 (and also addressed #541). The CLI shell-out approach in this PR was preserved as a fallback in case libvips proved untenable, but with libvips landed and the cruise on the horizon, keeping a parallel approach open adds noise without value. The branch on my fork remains in case anyone needs to reference the CLI logic later. Thanks all for the review feedback that pointed toward the better solution.
